### PR TITLE
Extend the capability of `decode` RPC to decrypt the contents of emergency.recover file.

### PR DESCRIPTION
**Pull request by adi2011**
This RPC method would help users to verify if the backup stored in emergencyrecover and peer_storage is valid and up-to-data. As requested in issue #6711.

### CODE REVIEW

File: plugins/chanbackup.c [hunk 1]
```diff
@@ -760,6 +760,128 @@ static struct command_result *json_restorefrompeer(struct command *cmd,
...
 }
 
+/* Returns True if the backup data is correct and
+ * returns false with the reason in `fail` if it's incorrect.
+ * Instead of using `decrypt_scb()` we're using this because
+ * it shuts down the node in case of any discrepancy and that
+ * is not what we want to happen when someone runs `verifybackup`. */
+static bool verify_bkp(struct command *cmd,
+					   u8 *filedata,
+					   char **fail) {
```
Comment (vincenzopalazzo) [inline plugins/chanbackup.c:770]: The indentation here look a little bit strange, the u8 should be in the same column of the struct

File: plugins/chanbackup.c [hunk 2]
```diff
@@ -760,6 +760,128 @@ static struct command_result *json_restorefrompeer(struct command *cmd,
...
+
+	crypto_secretstream_xchacha20poly1305_state crypto_state;
```
Comment (vincenzopalazzo) [inline plugins/chanbackup.c:772]: wow what name :)

There is any way to define a alias on top of the fine?

```
/* Re aliasing the crypto_secretstream_xchacha20poly1305_state type */
typedef crypto_secretstream_xchacha20poly1305_state bk_crypto_state;
```

File: plugins/chanbackup.c [hunk 3]
```diff
@@ -760,6 +760,128 @@ static struct command_result *json_restorefrompeer(struct command *cmd,
...
+	u64 version;
+	u32 timestamp;
+	struct scb_chan **scb;
+
+	if (tal_bytelen(filedata) < ABYTES +
+	    HEADER_LEN)
+	{
```
Comment (vincenzopalazzo) [inline plugins/chanbackup.c:779]:
```suggestion
	if (tal_bytelen(filedata) < 
	                ABYTES + HEADER_LEN) {
```

File: plugins/chanbackup.c [hunk 4]
```diff
@@ -760,6 +760,128 @@ static struct command_result *json_restorefrompeer(struct command *cmd,
...
+
+	u8 *decrypt_scb = tal_arr(tmpctx, u8, tal_bytelen(filedata) -
+				  ABYTES -
+				  HEADER_LEN);
+
+	/* The header part */
+	if (crypto_secretstream_xchacha20poly1305_init_pull(&crypto_state,
+							    filedata,
+							    (&secret)->data) != 0)
+	{
```
Comment (vincenzopalazzo) [inline plugins/chanbackup.c:792]: only the function has the starting block (`{`) in the next line, the if and other statement no

File: plugins/chanbackup.c [hunk 5]
```diff
@@ -760,6 +760,128 @@ static struct command_result *json_restorefrompeer(struct command *cmd,
...
+						       NULL, 0) != 0) {
+		*fail =  "data is corrupted!";
+		return false;
+	}
+
+	if (!fromwire_static_chan_backup(cmd,
+									 decrypt_scb,
+									 &version,
+									 &timestamp,
+									 &scb)) {
```
Comment (vincenzopalazzo) [inline plugins/chanbackup.c:812]: tabs looks stage

File: plugins/chanbackup.c [hunk 6]
```diff
@@ -760,6 +760,128 @@ static struct command_result *json_restorefrompeer(struct command *cmd,
...
+		*fail = "Incompatible emergency.recover file version on disk,"
+				"contact the admin or update the file asap!";
+		return false;
+	}
+	return true;
+}
+
+static struct command_result *after_fetchscb(struct command *cmd,
+						       const u8 *hexdata,
+						       struct json_stream *response)
```
Comment (vincenzopalazzo) [inline plugins/chanbackup.c:827]: strange identification

File: plugins/chanbackup.c [hunk 7]
```diff
@@ -760,6 +760,128 @@ static struct command_result *json_restorefrompeer(struct command *cmd,
...
+	struct scb_chan **scb;
+
+	if (tal_bytelen(hexdata) == 0) {
+		json_add_string(response, "peer_storage",
+				"Does not exist!");
+		return command_finished(cmd, response);
+	}
+
+	if (!fromwire_static_chan_backup(cmd,
+									 hexdata,
```
Comment (vincenzopalazzo) [inline plugins/chanbackup.c:840]: same here and in possible other places down there

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
Looks good to me.

We should add some docs about these commands, and I think also 
some use cases when the user should use the `emergency.recover` in general.

Are you planning to do something like this?

File: plugins/chanbackup.c (hunk 1 above)
Comment (adi2011) [reply plugins/chanbackup.c:770]: Hmm, my editor's setting is all over the place, actually this is a different machine. Fixing in a bit, thanks for pointing out!

File: plugins/chanbackup.c (hunk 2 above)
Comment (adi2011) [reply plugins/chanbackup.c]: Nice idea, Thanks!

File: plugins/chanbackup.c (hunk 3 above)
Comment (adi2011) [reply plugins/chanbackup.c]: fixed!

File: plugins/chanbackup.c (hunk 4 above)
Comment (adi2011) [reply plugins/chanbackup.c]: Done

File: plugins/chanbackup.c (hunk 5 above)
Comment (adi2011) [reply plugins/chanbackup.c]: Yes, resolved!

File: plugins/chanbackup.c (hunk 6 above)
Comment (adi2011) [reply plugins/chanbackup.c]: Resolved!

**Review by rustyrussell — COMMENTED** [review]
Minor changes only...

File: plugins/offers.c [hunk 8]
```diff
@@ -159,8 +164,33 @@ struct decodable {
...
+	u8 *encrypted_blob;
 };
 
+static u8 *encrypted_decode(const tal_t *ctx, const char *str, char **fail) {
+	size_t hrp_maxlen = strlen(str) - 6;
+	char *hrp = tal_arr(ctx, char, hrp_maxlen);
+
+	size_t data_maxlen = strlen(str) - 8;
+	u5 *data = tal_arr(ctx, u5, data_maxlen);
+	size_t datalen = 0;
```
[the remark this answers is not in the document]
Comment (adi2011) [reply plugins/offers.c]: Thank you for pointing this out!

File: plugins/offers.c [hunk 9]
```diff
@@ -1009,6 +1050,50 @@ static void json_add_rune(struct command *cmd, struct json_stream *js, const str
...
+	crypto_secretstream_xchacha20poly1305_state crypto_state;
+
+	u8 *decrypt_blob = tal_arr(tmpctx, u8,
+				   tal_bytelen(decodable->encrypted_blob) -
+				   ABYTES -
+				   HEADER_LEN);
+	/* The header part */
+	if (crypto_secretstream_xchacha20poly1305_init_pull(&crypto_state,
+							    decodable->encrypted_blob,
+							    (&secret)->data) != 0) {
```
[the remark this answers is not in the document]
Comment (adi2011) [reply plugins/offers.c]: agreed.

File: plugins/offers.c [hunk 10]
```diff
@@ -1036,6 +1121,21 @@ static struct command_result *json_decode(struct command *cmd,
...
+		const char *info = "scb secret";
+		u8 *info_hex = tal_dup_arr(tmpctx, u8,
+					   (u8*)info,
+					   strlen(info), 0);
+
+		req = jsonrpc_request_start(cmd->plugin, cmd, "makesecret",
+					    after_makesecret, &forward_error,
+					    decodable);
+
+		json_add_hex(req->js, "hex", info_hex, tal_bytelen(info_hex));
```
[the remark this answers is not in the document]
Comment (adi2011) [reply plugins/offers.c]: Yes, my bad. Thanks!

File: doc/lightning-decode.7.md [hunk 11]
```diff
@@ -15,6 +15,7 @@ The **decode** RPC command checks and parses:
   or `LIGHTNING:`) as specified by the BOLT 11 and BOLT 12
   specifications.
 - a *rune* as created by lightning-commando-rune(7).
+- a *encrypted\_blob* which contains the content of emergency.recover file.
```
[the remark this answers is not in the document]
Comment (adi2011) [reply doc/lightning-decode.7.md]: Yes, it's always better to give out more info.

**Review by rustyrussell — APPROVED** [review]
I'll just add some Changelog-Added lines, and squash regen commits (I need to rebase on master now, since other things got merged)

**Comment by rustyrussell** [comment]
Oh, and I changed "encrypted blob" from `decode` to "emergency recover" which is more specific!

**Comment by adi2011** [comment]
Great, Just needed to change the `type` given out in the result. Hopefully CI would we happy now.

File: plugins/offers.c [hunk 12]
```diff
@@ -159,8 +164,38 @@ struct decodable {
...
+
+	u8 *data8bit = tal_arr(data, u8, 0);
+	bech32_pull_bits(&data8bit, data, datalen*5);
+
+	return data8bit;
+fail:
+	tal_free(data);
+	return NULL;
+}
+
```
Comment (rustyrussell) [inline plugins/offers.c:198]: You changed this to set `*fail`, but it doesn't in various paths?

**Review by rustyrussell — CHANGES_REQUESTED** [review]

File: plugins/offers.c (hunk 12 above)
Comment (adi2011) [reply plugins/offers.c]: Thanks for pointing out.

**Review by rustyrussell — APPROVED** [review]

**Pull request merged**